### DISCUSSION TITLE
Help writing data to SQL

### DISCUSSION THREAD

**Discussion by legatrix** [jspsych/jsPsych#856] [Q&A] [answered]
I cannot write the results of the n-back task (YouTube tutorial video) to my SQL database. I am new to MySQL but I am sure the database I have created has a table with the right name and the right number of columns, named using `jsPsych.data.get().uniqueNames()`. I filled in all the SQL user / password / name fields carefully in my `database_config.php` file, using `localhost` and port 3306. Neither of the ways to write to an SQL database suggested on the jsPsych reference pages worked for me, so I found my way to the helper library

https://github.com/vassar-cogscilab/php-mysql-save-data

but sadly that method did not work either. With the latter method, the console returns nothing after the experiment completes. With the former methods, the console returns `undefined`. The actual data are fine, and accessible from the console after the experiment completes. Are there any common problems or oversights that may be hobbling me? Thank you in advance.

**Comment by jodeleeuw** [comment]
Hi @legatrix ,

Are you running on a live server or a local development server at this point? Are you able to login to the database and view the tables OK? 

I use the helper library you've found all the time, so I'm confident that it works when everything is configured correctly. 

What does your save data JS code look like?

**Reply by legatrix** [reply]
Hi Josh,

Thanks for getting back to me. This is on a live server provided by the shared hosting service (Namecheap) for my website. I am using phpMyAdmin through CPanel to create the database and the tables it contains. I can log in and view the tables OK:

<img width="1124" alt="Screenshot 2020-07-13 at 22 37 50" src="https://user-images.githubusercontent.com/8539784/87351331-9291a080-c559-11ea-8ffd-c720d822ac50.png">

The save data code is as recommended in the php-mysql-save-data documentation (just before `jsPsych.init()`):

```
<!DOCTYPE html>
<html>
	<head>
		<script src = "jsPsych-6.1.0/jspsych.js"></script>
		<script src = "jsPsych-6.1.0/plugins/jspsych-html-keyboard-response.js"></script>
		<script src = "jsPsych-6.1.0/plugins/jspsych-call-function.js"></script>
		<script src="js/serverComm.js" type="text/javascript"></script>
		<link href = "jsPsych-6.1.0/css/jspsych.css" rel = "stylesheet" type = "text/css"></link>
        </head>
		<body>
		</body>
		<script>

			var timeline = [];

			var n_back_set = ['Z', 'X', 'C', 'V', 'B', 'N'];
			var sequence = [];

			var n_back_trial = {
				type: "html-keyboard-response",
				stimulus: function(){
					if(sequence.length < 2){
						var letter = jsPsych.randomization.sampleWithoutReplacement(n_back_set,1)[0]
					} else{
						if(jsPsych.timelineVariable('match',true) == true){
							var letter = sequence[sequence.length - 2]
						} else{
							var possible_letters = jsPsych.randomization.sampleWithoutReplacement(n_back_set,2)
							if(possible_letters[0] != sequence[sequence.length - 2]){
								var letter = possible_letters[0]
							} else{
								var letter = possible_letters[1]
							}
						}
					}
					sequence.push(letter);
					return '<span style="font-size: 96px;">'+letter+'</span>'
				},
				choices: ['M'],
				trial_duration: 1500,
				response_ends_trial: false,
				post_trial_gap: 500,
				// 'match' won't make sense for the first two trials...
				// de Leeuw mentioned this later.
				data: {
					match: jsPsych.timelineVariable('match')
				},
				on_finish: function(data){
					if(data.match == true){
						data.correct = (data.key_press != null)
					}
					if(data.match == false){
						data.correct = (data.key_press === null)
					}
				}
			}

			// timeline.push(n_back_trial)

			var n_back_trials = [
				{match: true},
				{match: false}
			]

			var n_back_sequence = {
				timeline: [n_back_trial],
				timeline_variables: n_back_trials,
				sample: {
					type: 'with-replacement',
					size: 4,
					weights: [1,2]
				}
			}

			timeline.push(n_back_sequence)
			
			var save = {
              type: 'call-function',
              func: function(){
                serverComm.save_data(jsPsych.data.get().values());
              }
            }
            
            timeline.push(save);

			jsPsych.init({
				timeline: timeline
			})
		</script>
</html>
```

I am getting an error message which seems to suggest that SQL is at least being accessed:

`error_message: "SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '-backResults VALUES(NULL, '<span style=\\\"font-size: 96px;\\\">B</span>', NULL, ...' at line 1"`

Sadly I am too incompetent to figure out how to view the truncated part of the error message in Firefox or Safari!

(Also, the data saves fine to the server with the simpler .csv file method, so it is fine in principle to write to the server.)

Thanks,

Luke

**Comment by legatrix** [comment, marked as answer]
Dear Josh, 

I am happy to report that I finally diagnosed the problem! By looking at the error message in the last post, I noticed the "n" was missing from "n-backResults", which led me to rename everything without the hyphen, and it worked. I guess SQL doesn't like hyphens!

**Reply by jodeleeuw** [reply]
Interesting! Was that the name of your table?

**Reply by legatrix** [reply]
The former name was `n-backResults` and the new name is `nbackResults`.
